vertx-mongo-client/src/main/java/io/vertx/ext/mongo/impl/MongoIterableStream.java

```diff
@@ -0,0 +1,186 @@
+package io.vertx.ext.mongo.impl;
+
+import com.mongodb.async.AsyncBatchCursor;
+import com.mongodb.async.SingleResultCallback;
+import com.mongodb.async.client.MongoIterable;
+import io.vertx.core.Context;
+import io.vertx.core.Handler;
+import io.vertx.core.json.JsonObject;
+import io.vertx.core.streams.ReadStream;
+
+import java.util.ArrayDeque;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * @author Thomas Segismont
+ */
+class MongoIterableStream implements ReadStream<JsonObject> {
+
+  private static final int BATCH_SIZE = 10;
+
+  private final Context context;
+  private final MongoIterable<JsonObject> mongoIterable;
+
+  private AsyncBatchCursor<JsonObject> batchCursor;
+  private Deque<JsonObject> queue;
+  private Handler<JsonObject> dataHandler;
+  private Handler<Throwable> exceptionHandler;
+  private Handler<Void> endHandler;
+  private boolean paused;
+  private boolean readInProgress;
+  private boolean closed;
+
+  MongoIterableStream(Context context, MongoIterable<JsonObject> mongoIterable) {
+    this.context = context;
+    this.mongoIterable = mongoIterable;
+  }
+
+  @Override
+  public synchronized MongoIterableStream exceptionHandler(Handler<Throwable> handler) {
+    checkClosed();
+    this.exceptionHandler = handler;
+    return this;
+  }
+
+  private void checkClosed() {
+    if (closed) {
+      throw new IllegalArgumentException("Stream is closed");
+    }
+  }
+
+  @Override
+  public synchronized MongoIterableStream handler(Handler<JsonObject> handler) {
+    checkClosed();
+    if (handler == null) {
+      close();
+    } else {
+      dataHandler = handler;
+      SingleResultCallback<AsyncBatchCursor<JsonObject>> callback = (result, t) -> {
+        context.runOnContext(v -> {
+          synchronized (this) {
+            if (t != null) {
+              close();
+              handleException(t);
+            } else {
+              batchCursor = result;
+              if (canRead()) {
+                doRead();
+              }
+            }
+          }
+        });
+      };
+      try {
+        mongoIterable.batchCursor(callback);
+      } catch (Exception e) {
+        close();
+        handleException(e);
+      }
+    }
+    return this;
+  }
+
+  private boolean canRead() {
+    return !paused && !closed;
```

Comment:
Same comment as above about volatile and synchronized.

Reply:
Also guarded by the stream instance

Reply:
comment this again for follow on developers